src/deep_sea_vectorization.rs

```diff
@@ -0,0 +1,414 @@
+        let position_val: u16 = match self {
+            Position::Diving(depth) => *depth as u16,
+            _ => 0,
+        };
+        [T::from(position_val),
+         T::from(*self == Position::WaitingToDive),
+         T::from(*self == Position::ReturnedToSubmarine)].into_iter()
+    }
+
+    fn unpacked_size(&self) -> usize {
+        POSITION_ENUM_COUNT
+    }
+}
+
+
+impl Unpackable for Tile {
+    fn unpack<T: DataType>(&self) -> impl Iterator<Item = T> {
+        match self {
+            Tile::Empty => UnifiedIterator::Opt1([T::zero()].into_iter()),
+            Tile::Treasure(treasure) => UnifiedIterator::Opt2(treasure.unpack::<T>()),
+        }
+    }
+
+    fn unpacked_size(&self) -> usize {
+        1
+    }
+}
+
+
+
+impl Unpackable for Path {
+    fn unpack<T: DataType>(&self) -> impl Iterator<Item = T> {
+        let path_iter = self.tiles.unpack::<T>();
+        let occupied_iter = (0..self.tiles.len()).map(|idx| T::from(self.occupied.contains(idx)));
+        itertools::interleave(path_iter, occupied_iter)
+    }
+
+    fn unpacked_size(&self) -> usize {
+        2 * self.tiles.len()
+    }
+}
+
+
+impl Unpackable for Player {
+    fn unpack<T: DataType>(&self) -> impl Iterator<Item = T> {
+        let direction = self.direction.unpack::<T>();
+        let position = self.position.unpack::<T>();
+        let held_treasures = self.held_treasures.unpack::<T>();
+        let filled_size = self.direction.unpacked_size()
+            + self.position.unpacked_size()
+            + self.held_treasures.unpacked_size();
+        direction.chain(position)
+                 .chain(held_treasures)
+                 .chain(std::iter::repeat_n(
+                     T::zero(),
+                     self.unpacked_size().saturating_sub(filled_size)))
+    }
+
+    fn unpacked_size(&self) -> usize {
+        self.direction.unpacked_size() + self.position.unpacked_size() + MAX_NUM_TREASURES
+    }
+}
+
+
+
+impl Unpackable for DeepSeaAction {
+    fn unpack<T: DataType>(&self) -> impl Iterator<Item = T> {
+        // let mut unpacked = vec![T::zero(); DEEP_SEA_ACTION_ENUM_COUNT];
+        let hot_idx = match self {
+            DeepSeaAction::TreasureDecision(td) => match td {
+                TreasureDecision::Ignore => 0,
+                TreasureDecision::Take => 1,
+                TreasureDecision::Return(t) => 2 + *t as usize,
+            },
+            DeepSeaAction::DiveDirection(dd) => TREASURE_DECISION_ENUM_COUNT + *dd as usize,
+        };
+        std::iter::repeat_n(T::zero(), hot_idx)
+            .chain(std::iter::once(T::from(1u16)))
+            .chain(std::iter::repeat_n(
+                T::zero(), self.unpacked_size() - hot_idx - 1))
+    }
+
+    fn unpacked_size(&self) -> usize {
+        DEEP_SEA_ACTION_ENUM_COUNT
+    }
+}
+
+
+impl Unpackable for DeepSeaState {
+    fn unpack<T: DataType>(&self) -> impl Iterator<Item = T> {
+        self.path.unpack::<T>().collect::<Vec<T>>().into_iter()
```

Comment:
```suggestion
        self.path.unpack::<T>().collect_vec().into_iter()
```
(itertools convenience function)

Reply:
Done